flo-workflow/src/main/java/com/spotify/flo/EvalContext.java

```diff
@@ -67,8 +71,42 @@
    * @param <T>      The type of the task result
    * @return A value of the task result
    */
+  @SuppressWarnings("unchecked")
   default <T> Value<T> evaluateInternal(Task<T> task, EvalContext context) {
-    return task.code().eval(context);
+    // Get the process fn input arguments
+    final Value<List<Object>> args = task.args().stream()
+        .map(arg -> arg.get(context))
+        .map(value -> value.map(v -> (Object) v))
+        .collect(Values.toValueList(context));
+
+    return args.flatMap(as -> {
+
+      // Call preRun on task contexts after all args are evaluated
+      task.contexts().forEach(tc -> tc.preRun(task));
+
+      final Optional<TaskOperator> operator = task.contexts().stream()
+          .filter(c -> c instanceof TaskOperator)
+          .map(c -> (TaskOperator) c)
+          .findAny();
+
+      // Run the process fn
+      final Listener listener = listener();
+      final Invokable processFn = task.processFn();
+      final Value<Object> value = context.invokeProcessFn(task.id(), () ->  {
+        final Object result = processFn.invoke(as.toArray());
+
+        // Run operator
+        return operator
+            .map(o -> o.perform(result, listener))
+            .orElse(result);
```

Comment:
Does this rely on type checks from `TaskBuilder`? If an operator exists, then `result` will be the same as the `T` type specified by `TaskOperator<T, Y, Z>` and if no operator exists this will be the `T` specified by `Task<T>`? If so, a comment mentioning that would be good here.

Reply:
Yeah, we rely on type checking happening in the `TaskBuilder` interface.

If an operator exists, result will be `Y` in `TaskOperator<T, Y, Z>`. Will add a comment.

Maybe we should have better names for the type arguments =)

Reply:
I was thinking that too now with these types that actually mean something within flo.